### ISSUE TITLE
Terraform init with Azure DevOps and https for module source

### ISSUE THREAD

**Issue by Joerg-L** [hashicorp/setup-terraform#223]
So, we are using Azure Devops to store our Terraform config and all the self created module code.
We also want to use a DevOps Pipeline to apply the configuration.

As we are not allowed to use ssh for accessing the repos (traffic has to go trough the https-proxy), so we have to use https for the repository integration so that our source for the module looks like
```
 source = "git::https://<<ADO_ORG>>@dev.azure.com/<ADO_ORG>>/<<ADO_PROJECT>>>/_git/<<ADO_REPO>>"
```

Locally running `terraform init` works completely fine. When running it in the pipeline we see following

```
│ Error: Failed to download module
│ 
│ Could not download module "xyz" (main.tf:3)
│ source code from
│ "git::https://<<ADO_ORG>>@dev.azure.com/<ADO_ORG>>/<<ADO_PROJECT>>>/_git/<<ADO_REPO>>"
│ error downloading
│ 'https://<<ADO_ORG>>@dev.azure.com/<ADO_ORG>>/<<ADO_PROJECT>>>/_git/<<ADO_REPO>>'
│ /usr/bin/git exited with 128: Cloning into
│ '.terraform/modules/xyz'...
│ fatal: could not read Password for 'https://<<ADO_ORG>>@dev.azure.com':
│ terminal prompts disabled
```

We have tested many things right now and only with changeing source to
```
source = "git::https://<<PAT>>@dev.azure.com/<ADO_ORG>>/<<ADO_PROJECT>>>/_git/<<ADO_REPO>>"
```

we were able to run `terraform init` but checking in PAT to git sounds not very right.

Anyone an idea on that?
Regards
Joerg

**Comment by VOVELEE** [comment]
I am hitting the same issue. Anyone with ideas?

**Comment by srjennings** [comment]
Same issue.

**Comment by Joerg-L** [comment]
> Get the same issue, just found a workaround with the system access token.
> 
thanks @b1n9s  I will take a look at that workaround.

It looks usefull if there is only one (or a few) repository(s), but we have for each module a single repository, so as of now 27 module repos.

**Comment by Joerg-L** [comment]
@b1n9s so, I have tested the workarround

I isn't working as we have more then 20 module repos and azure is limiting that to max 20.

**Comment by Joerg-L** [comment]
@srjennings @VOVELEE @b1n9s 
so, after many tests I was able to solve the issue now. As I can see, the solution has two parts.

First part of our solution:
Script in the pipeline yml
```
- script: |
    git config --global url."https://$SYSTEM_ACCESSTOKEN@dev.azure.com".insteadOf "https://<<ADO-ORG>>@dev.azure.com"
  displayName: 'set extra header'
  env:
    SYSTEM_ACCESSTOKEN: $(System.AccessToken)
 ```

that is what also @b1n9s has mentioned.
But, as mentioned, adding the repos as resources isn't an option for us because of the number of repos and, what is more important, that option is limited to 20 repos in total.

Second part of our solution:
Deaktivating the setting "Protect access to repositories in YAML pipelines"

With both parts together, everything seams to work now.

**Comment by IanMoroney** [comment]
Can confirm, @Joerg-L 's solution works!

**Comment by mldevpants** [comment]
> Get the same issue, just found a workaround with the system access token.
> 
>     * Step 1, config your terraform module [repository resource](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema#define-a-repositories-resource) in the pipeline
> 
> 
> ```yaml
> resources:
>   repositories: 
>     - repository: terraform-modules
>       type: git
>       name: terraform-modules
> ```
> 
>     * Step 2: add the following step before your `terraform init`, change _<your_org>_ to your organization name
> 
> 
> ```yaml
> steps:
>   - bash: git config --global http.https://<your_org>@dev.azure.com.extraheader "AUTHORIZATION: bearer $(System.AccessToken)"
> ```
> 
> And then you should be able to run `terraform init`

Wow, I was hitting my had to the wall with this one, I am running a terragrunt with azure repos as git, using it inside the container job, and I tried a lot of things inlcuding using the ssh key as git_ssh_command, tried to install using the installsshkey, with this I switched the sources to use https rather than ssh, and with this little trick it finally worked.
I have similar jobs with with only terraform, no terragrunt, everything works on that part but with terragrunt a new issue.
Thank you

**Comment by Joerg-L** [comment]
Hey @Isaac-Irvine 
> This works for me when getting repos in the same project but fails when I try to get repos from other projects. Does anyone have any idea how to fix that?

The problem is that the System.AccessToken is bounded to the project specific build account.

I have just checkt it in Azure DevOps, you can grant permissions in "ProjectB" to the Build Account of "ProjectA".
So I could imagine that if you update the permissions, it could work.

**Comment by guidooliveira** [comment]
> > @srjennings @VOVELEE @b1n9s so, after many tests I was able to solve the issue now. As I can see, the solution has two parts.
> > First part of our solution: Script in the pipeline yml
> > ```
> > - script: |
> >     git config --global url."https://$SYSTEM_ACCESSTOKEN@dev.azure.com".insteadOf "https://<<ADO-ORG>>@dev.azure.com"
> >   displayName: 'set extra header'
> >   env:
> >     SYSTEM_ACCESSTOKEN: $(System.AccessToken)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > that is what also @b1n9s has mentioned. But, as mentioned, adding the repos as resources isn't an option for us because of the number of repos and, what is more important, that option is limited to 20 repos in total.
> > Second part of our solution: Deactivating the setting "Protect access to repositories in YAML pipelines" ![image](https://user-images.githubusercontent.com/25051496/194055291-dc1bd1a7-dd3d-41ee-bba3-b20cc56bcbfb.png)
> > With both parts together, everything seams to work now.
> 
> Thanks for that! This works for me when getting repos in the same project but fails when I try to get repos from other projects. Does anyone have any idea how to fix that?

In case anyone is still having issues, I found that using http.<CollectionUri>.extraheader to work best for inserting and cleaning up afterwards as it keeps the token part of the value, as opposed to the URI.
insert this prior to terraform init
```yaml
- pwsh: git config --global --add http.${{variables['System.CollectionUri']}}.extraheader "AUTHORIZATION: bearer $(System.AccessToken)"
```

cleanup task:
```yaml
- pwsh: git config --global --unset-all http.${{variables['System.CollectionUri']}}.extraheader 
```

**Comment by tyagivasu** [comment]
Hello, All can someone please help me with the sourcing of specific modules...

Currently, I find the below solution which works for me locally but I would like to know if there is a better approach someone follows.

```tf
module "my-module" {
  source = "git::https://dev.azure.com/<org>/<project>/_git/<repo>//<module-folder>?ref=<branch||tag>"
  variable = ...
  variable = ...
}
```
++ @Joerg-L 
Thanks in advance :)

**Comment by b1n9s** [comment]
> 
@tyagivasu 
you could also use multi-repo checkout and [specify ref in the pipeline resources](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema#define-a-repositories-resource). 

But then you would need to use a relative path for sourcing in the module, not sure if it's a better approach for you.

**Comment by tyagivasu** [comment]
> > 
> 
> @tyagivasu you could also use multi-repo checkout and [specify ref in the pipeline resources](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema#define-a-repositories-resource).
> 
> But then you would need to use a relative path for sourcing in the module, not sure if it's a better approach for you.

Hi @b1n9s thanks for your reply. Actually, that is what exactly I have today :) But I am looking to move in the HTTP:// source direction to roll out specific module versions, And would like to hear from people on any challenges, and approaches are taken while going to the https:// kind of sourcing.

**Comment by gregdskb** [comment]
> @srjennings @VOVELEE @b1n9s so, after many tests I was able to solve the issue now. As I can see, the solution has two parts.
> 
> First part of our solution: Script in the pipeline yml
> 
> ```
> - script: |
>     git config --global url."https://$SYSTEM_ACCESSTOKEN@dev.azure.com".insteadOf "https://<<ADO-ORG>>@dev.azure.com"
>   displayName: 'set extra header'
>   env:
>     SYSTEM_ACCESSTOKEN: $(System.AccessToken)
> ```
> 
> that is what also @b1n9s has mentioned. But, as mentioned, adding the repos as resources isn't an option for us because of the number of repos and, what is more important, that option is limited to 20 repos in total.
> 
> Second part of our solution: Deactivating the setting "Protect access to repositories in YAML pipelines" ![image](https://user-images.githubusercontent.com/25051496/194055291-dc1bd1a7-dd3d-41ee-bba3-b20cc56bcbfb.png)
> 
> With both parts together, everything seams to work now.

Thanks @Joerg-L

**Comment by bmv0161** [comment]
I'm facing a similar issue for a github repo. Have not had success following these solutions. Has anyone else faced this?

**Comment by gregdskb** [comment]
Yes @bmv0161. I recently ported from ADO to GitHub too. Similar issues

**Comment by b1n9s** [comment]
@bmv0161 @gregdskb can you give more details about the setup, is it one of the following?

1. Caller repo and the module repo are both on Github, or
2. Caller repo on Github and module repo on ADO, or 
3. Caller repo on ADO and module repo on Github

**Comment by gregdskb** [comment]
> @bmv0161 @gregdskb can you give more details about the setup, is it one of the following?
> 
> 1. Caller repo and the module repo are both on Github, or
> 2. Caller repo on Github and module repo on ADO, or
> 3. Caller repo on ADO and module repo on Github

For me @b1n9s , it was caller repo in ADO and module repo in ADO. Then migrated my ADO pipeline to GitHub Actions. Caller repo in GitHub and module repo in GitHub. Both had similar issues. Not sure this is a TF issue. Perhaps better discussed on Stackoverflow

**Comment by b1n9s** [comment]
> For me @b1n9s , it was caller repo in ADO and module repo in ADO. Then migrated my ADO pipeline to GitHub Actions. Caller repo in GitHub and module repo in GitHub. Both had similar issues. Not sure this is a TF issue. Perhaps better discussed on Stackoverflow

@gregdskb  For sure this is not a TF issue.

Not sure how your auth is configured in the GitHub Actions, but one thing to note, the `GITHUB_TOKEN` secret has limited access to its own repo only if that's what you are using. https://docs.github.com/en/actions/security-guides/automatic-token-authentication
```
The token's permissions are limited to the repository that contains your workflow.
```

**Comment by gregdskb** [comment]
Thanks @b1n9s , using the `GITHUB_TOKEN` is not the answer. If you or @bmv0161 want to start a topic in Stackoverflow I will contribute my solutions. Hope that helps

**Comment by bmv0161** [comment]
> @bmv0161 @gregdskb can you give more details about the setup, is it one of the following?
> 
> 1. Caller repo and the module repo are both on Github, or
> 2. Caller repo on Github and module repo on ADO, or
> 3. Caller repo on ADO and module repo on Github

For me, both the caller repo and the module repo are in Github. With deployment in Azure pipelines. Attempting to use the System.AccessToken with a Github Apps Service Connection does not seem to authenticate properly.

**Comment by alexanderpilch** [comment]
Here's another solution which is using environment variables instead. You won't have to use "global variables" so there's no risk of issues with multiple builds in parallel (and leaking tokens to other builds). 
`git remote -v` is also listing the original url after the build has completed as the override is gone. The access token is not included in there anymore.

Azure pipeline:
```yaml
steps:
  - script: |
      export GIT_CONFIG_PARAMETERS="'url.https://$SYSTEM_ACCESSTOKEN@dev.azure.com.insteadof=https://<YOUR_ORG>@dev.azure.com'" 
      terraform init
    env:
      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
    displayName: Initialize Terraform modules
````

The module is included via this snippet:
```terraform
module "mymodule" {
  source = "git::https://<YOUR_ORG>@dev.azure.com/<YOUR_ORG>/<YOUR_PROJECT>/_git/<YOUR_REPO>//<PATH_IN_REPO>"
}
```

**Comment by burizz** [comment]
hitting the same issue and as mentioned by @karts499  I tried all options suggested in comments here and still getting the same problem : 

![image](https://github.com/hashicorp/setup-terraform/assets/5168054/14b551b6-318e-4e1c-9c1a-645839067978)

**Comment by byrontuckett** [comment]
So I'm not a fan of disabling this project wide setting __Protect access to repositories in YAML pipelines__. I don't think that is a good idea.  [Read more about job access tokens here](https://learn.microsoft.com/en-us/azure/devops/pipelines/process/access-tokens?view=azure-devops&tabs=yaml#protect-access-to-repositories-in-yaml-pipelines). There is a reason it is enabled by default.

With a test configuration repo and a test terraform modules repo hosted in the same Project I managed to get it working.
Couple of things:
* I only tested with MSFT hosted agents.  Did not test on Self Hosted.
* Used a single repository for all terraform modules.  Don't believe this would be suitable for a high number of repositories due to AzDo limits.

First adding the additional repository as a resource and explicitly in a checkout step.
```yaml
resources:
 repositories:
   - repository: terraform_modules
     type: git
     name: terraform_modules

steps:
- checkout: self
- checkout: terraform_modules
```

When the pipeline runs you will be prompted to authenticate to the extra repository (hint, job access token now has permission to read from that repo 🤘)

Add the git extraHeader config. I chose to use an environment variable which will only last per the variable scope (the job in the case of my test pipeline).  If you choose to use/set git global config (especially on a self hosted agent) there may be danger ahead if the agent is shared etc.

```yaml
- script: |
    export GIT_CONFIG_PARAMETERS="'url.https://$SYSTEM_ACCESSTOKEN@dev.azure.com.insteadof=https://<YOUR_ORG_NAME_GOES_HERE>@dev.azure.com'" 
    terraform --version -json
    terraform init
    terraform plan
  env:
    SYSTEM_ACCESSTOKEN: $(System.AccessToken)
  displayName: Test Init and Plan
  workingDirectory: $(Build.SourcesDirectory)/config_test_repo/test1
```

Really that is it.

Example of a module source:
```terraform

module "name" {
  source = "git::https://<ORGNAME>@dev.azure.com/<ORGNAME>/<PROJECT_NAME>/_git/terraform_modules//modules/<MODULE_NAME>?ref=v2"
  ...
}
```

Success. Can use `git:https` with the system access token.

<img width="963" alt="Screenshot 2024-04-11 at 12 31 12 am" src="https://github.com/hashicorp/setup-terraform/assets/5583791/161858b2-fb23-4148-83b3-d60f829006d6">


#### References:

https://learn.microsoft.com/en-us/azure/devops/pipelines/security/secure-access-to-repos?view=azure-devops&tabs=yaml
https://learn.microsoft.com/en-us/azure/devops/pipelines/process/access-tokens?view=azure-devops&tabs=yaml#protect-access-to-repositories-in-yaml-pipelines
https://learn.microsoft.com/en-us/azure/devops/pipelines/repos/multi-repo-checkout?view=azure-devops#repository-resource-definition

**Comment by karts499** [comment]
Hey everyone. After long nights of debugging and try and errors. We finally found a workaround.
Just by simply using a bash task instead of a pwsh task:

` - bash: |
                find $(Build.SourcesDirectory)/ -type f -name 'main.tf' -exec sed -i 's~git::https://{ORG_NAME}@dev.azure.com~git::https://$(System.AccessToken)@dev.azure.com~g' {} \;`

I should basicly do the same like my try with pwsh. Dont understand what exactly made the difference.

Let me know if you have a better solution :)

**Comment by kollipara** [comment]
> Get the same issue, just found a workaround with the system access token.
> 
> * Step 1, config your terraform module [repository resource](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops&tabs=schema#define-a-repositories-resource) in the pipeline
> 
> ```yaml
> resources:
>   repositories: 
>     - repository: terraform-modules
>       type: git
>       name: terraform-modules
> ```
> 
> * Step 2: add the following step before your `terraform init`, change _<your_org>_ to your organization name
> 
> ```yaml
> steps:
>   - bash: git config --global http.https://<your_org>@dev.azure.com.extraheader "AUTHORIZATION: bearer $(System.AccessToken)"
> ```
> 
> And then you should be able to run `terraform init`

I was able to get it working with these steps + 1 additional step where I had to map the additional repositories that each job consumes.
```yaml
jobs:
- job:
   uses:
       repositories: 
       - terraform-modules
```
Reference:  https://learn.microsoft.com/en-us/azure/devops/pipelines/repos/azure-repos-git?view=azure-devops&tabs=yaml#protect-access-to-repositories-in-yaml-pipelines

**Comment by peter-mogaka** [comment]
Do the following steps:

1: Create a service connection for inter-repo access( option called "Azure Repos/Team Foundation Server"_). You'll need to create a PAT. 

2. Add your Remote project in the security section of the service connection. Click the service connection, then 3 dots(usually on the  top right side), then security, then project permissions.

3. Add your remote repo as a resource in your pipeline.

 repositories:
  - repository: your_remote_repo
    type: git
    name: <your_remote_project>/<your_remote_repo>
    endpoint: <service connection created in step 1>

5. Check out self and checkout your remote repo. **Important to set workspaceRepo**. It ensure that terraform inits in the current repo(not remote repo)

- checkout: self
   workspaceRepo: true  #must be included
 - checkout: your_remote_repo
    submodules: true    #only if you want submodules.


6.  Call module with "../" since it's checked to a directory on the runner.

module "<name>" {
  source = "../<your_remote_project>/<path_to_repo>"


P.S
**I'm using windows server 2022 runners. My repos are in separate projects.**

**Comment by K8sexpert** [comment]
Am also getting same issue, any solution ?

**Comment by karts499** [comment]
> Hey everyone. I am facing the same issue right now. We are trying to use a dedicated Azure Project for terraform modules. Within that, every module gets a single repo. like [@Joerg-L](https://github.com/Joerg-L) mentioned above we are trying to refrence each modul repo in the terraform configuration and useing the System.AccessToken in Pipeline as shown :
> 
> ```
> module "NAME" {
>   source    = "git::https://<ORGA>@dev.azure.com/<ORGA>/<PROJECT>/_git/<REPO>?ref=<BRANCH>"
> }
> ```
> 
> unfortunately none of the approaches working. Where are getting the following error message in the azure pipeline:
> 
> ```
> │ Error: Failed to download module
> │ 
> │   on main.tf line 26:
> │   26: module "MODULE" {
> │ 
> │ Could not download module "resources" (main.tf:26) source code from
> │ "git::[https://@dev.azure.com/<ORGA>/<PROJECT>/_git/.....
> │ error downloading
> │ 'https://<ORGA>@dev.azure.com/<ORGA>/<PROJECT>/_git/<REPO>?ref=<BRANCH>'
> (https://<ORGA>@dev.azure.com/<ORGA>/<PROJECT>/_git/<REPO>?ref=<BRANCH>%27):
> │ /usr/bin/git exited with 128: Cloning into
> │ '.terraform/modules/resources'...
> │ fatal: could not read Password for 'https://<ORGA>@dev.azure.com'/: terminal
> │ prompts disabled
> ```

Addition: make sure your pipelines settings are permissiv enough.

**Comment by byrontuckett** [comment]
Just wanted to check something as I see the same your error message. In my example, I used <ORG> <REPO> etc as placeholders so as not to expose my organisation name or project name into GitHub :) - be sure to swap them out for your actual organisation name and project name etc. 
I take it you are doing the same for this public issue and not actually leaving those placeholders in the code? Just wanted to make sure. 

I'll post a link to a repo showing the implementation tomorrow (sorry it's late here atm)

**Comment by K8sexpert** [comment]
@byrontuckett Any update please?

**Comment by byrontuckett** [comment]
@K8sexpert  So I got stuck with the [zero free cicd policy on new organisations](https://devblogs.microsoft.com/devops/change-in-azure-pipelines-grant-for-public-projects/?WT.mc_id=DOP-MVP-5001511) with public projects. So I'll just detail it here with screenshots and code blocks.

Lets just say my azure devops organisation name is: **acmecorp**  and my project name is **ai_agent_infra**.

In one repository I have `terraform_modules` repo.  Structure looks something like this:
```
  modules/example_1
    main.tf
    outputs.tf
    README.md
  modules/example_2
    ...and so on
```

Another repository houses the platform terraform files. Lets call this the `ai_agent_env_config` repository.
```
main.tf
README.md
...and so on
```

Now the `main.tf` file in the config repository has the module source as follows.  Here I'm using a tag of `v2` of the `example_1` module:
```tf
module "name" {
  source = "git::https://acemecorp@dev.azure.com/acemecorp/ai_agent_infra/_git/terraform_modules//modules/example_1?ref=v2"
  name     = "byron"
  location = "australiaeast"
}
```

Here is a simplified yaml pipeline that you can copy paste and adjust for your testing.

```yaml
trigger: none # Manually trigger the pipeline

pool:
  vmImage: 'ubuntu-latest'

 # https://learn.microsoft.com/en-us/azure/devops/pipelines/process/resources?view=azure-devops#repository-resource-definition
resources:
  repositories:
    - repository: terraform_modules
      type: git
      name: terraform_modules # The actual name of the repo

steps:
  - checkout: self # this will be ai_agent_env_config under the Build.SourcesDirectory path
  - checkout: terraform_modules

  - script: |
      echo "This step will continue on error and showcase a failing auth to our tf modules repo"
      terraform --version -json
      terraform init
      terraform plan
    displayName: Without AccessToken
    workingDirectory: $(Build.SourcesDirectory)/ai_agent_env_config # location of the root module/working tf directory
    continueOnError: true

  - script: |
      export GIT_CONFIG_PARAMETERS="'url.https://$SYSTEM_ACCESSTOKEN@dev.azure.com.insteadof=https://acemecorp@dev.azure.com'" 
      terraform --version -json
      terraform init
      terraform plan
    env:
      SYSTEM_ACCESSTOKEN: $(System.AccessToken)
    displayName: With AccessToken
    workingDirectory: $(Build.SourcesDirectory)/ai_agent_env_config # location of the root module/working tf directory
```

Key things to point out.

1. The `resources` property safely injects the repository into the scope of the pipeline access token.
2. In the second `script` step, Git is configured to use the System.AccessToken via a temporary environment variable `GIT_CONFIG_PARAMETERS`. This is only required for the `terraform init` step as we are using git to download modules from our private repository.
3. `SYSTEM_ACCESSTOKEN` must be injected into the step using the `env` yaml key as shown.

Screenshots of the pipeline in action:
This step is the fail on purpose step showing no access to the repository without the token.
![Image](https://github.com/user-attachments/assets/2c3c9e82-f9a2-45bb-b626-f316227c834e)

This step has the access token properly configured.
![Image](https://github.com/user-attachments/assets/403d1da3-2735-4271-9200-e28ec46844e8)

I hope this helps with your use case.  I've also been using this pattern on a self hosted windows agent setup (just with a few tweaks to make the variables work with PowerShell).

**Comment by K8sexpert** [comment]
@byrontuckett Thank you for your time :), I will check and update here.

**Comment by K8sexpert** [comment]
@byrontuckett I have tested your solution and its working. thanks :)

**Comment by craigthackerx** [comment]
> Hey @Isaac-Irvine
> 
> > This works for me when getting repos in the same project but fails when I try to get repos from other projects. Does anyone have any idea how to fix that?
> 
> The problem is that the System.AccessToken is bounded to the project specific build account.
> 
> I have just checkt it in Azure DevOps, you can grant permissions in "ProjectB" to the Build Account of "ProjectA". So I could imagine that if you update the permissions, it could work.

I'm actually struggling to get this working in my dev environment.  I have granted ProjectA's build service near enough full access to ProjectBs repos, as well as ProjectB to ProjectA's.

If anyone has a working solution where you can store terraform modules inside a separate _private_ Azure DevOps project, and use them on another project in that organisation, please post here or a link to a blog etc.

I have followed basically every option on this thread currently, disabled the "Protect" settings etc and still no luck.
